drivers/counter/counter_ifx_cat1.c

```diff
@@ -0,0 +1,488 @@
+	.period = 32768,
+	.clockPrescaler = CY_TCPWM_COUNTER_PRESCALER_DIVBY_1,
+	.runMode = CY_TCPWM_COUNTER_CONTINUOUS,
+	.countDirection = CY_TCPWM_COUNTER_COUNT_UP,
+	.compareOrCapture = CY_TCPWM_COUNTER_MODE_CAPTURE,
+	.compare0 = 16384,
+	.compare1 = 16384,
+	.enableCompareSwap = false,
+	.interruptSources = CY_TCPWM_INT_NONE,
+	.captureInputMode = 0x3U,
+	.captureInput = CY_TCPWM_INPUT_0,
+	.reloadInputMode = 0x3U,
+	.reloadInput = CY_TCPWM_INPUT_0,
+	.startInputMode = 0x3U,
+	.startInput = CY_TCPWM_INPUT_0,
+	.stopInputMode = 0x3U,
+	.stopInput = CY_TCPWM_INPUT_0,
+	.countInputMode = 0x3U,
+	.countInput = CY_TCPWM_INPUT_1,
+};
+
+static int32_t _get_hw_block_info(TCPWM_CNT_Type *reg_addr, cyhal_resource_inst_t *hw_resource)
+{
+	int32_t ret = -1;
+	uint32_t i;
+
+	for (i = 0u; i < _CYHAL_TCPWM_INSTANCES; i++) {
+		TCPWM_Type *base = _CYHAL_TCPWM_DATA[i].base;
+
+		if (((uint32_t)reg_addr > (uint32_t)base) &&
+		    ((uint32_t)reg_addr < (uint32_t)base + sizeof(TCPWM_Type))) {
+
+			hw_resource->type = CYHAL_RSC_TCPWM;
+			hw_resource->block_num = i;
+			hw_resource->channel_num = ((((uint32_t)reg_addr) - (uint32_t)base->CNT) /
+						    sizeof(TCPWM_CNT_Type));
+
+			if (hw_resource->channel_num < _CYHAL_TCPWM_DATA[i].num_channels) {
+				ret = 0;
+			}
+			break;
+		}
+	}
+
+	return ret;
+}
+
+static void ifx_cat1_counter_event_callback(void *callback_arg, cyhal_timer_event_t event)
+{
+	const struct device *dev = (const struct device *) callback_arg;
+	struct ifx_cat1_counter_data *const data = dev->data;
+	const struct ifx_cat1_counter_config *const config = dev->config;
+
+	/* Alarm compare/capture event */
+	if ((data->alarm_cfg_counter.callback != NULL) &&
+	    (((CYHAL_TIMER_IRQ_CAPTURE_COMPARE & event) == CYHAL_TIMER_IRQ_CAPTURE_COMPARE) ||
+	     data->alarm_irq_flag)) {
+		/* Alarm works as one-shot, so disable event */
+		cyhal_timer_enable_event(&data->counter_obj,
+					 CYHAL_TIMER_IRQ_CAPTURE_COMPARE,
+					 config->irq_priority, false);
+
+		/* Call User callback for Alarm */
+		data->alarm_cfg_counter.callback(dev, 1, cyhal_timer_read(&data->counter_obj),
+						 data->alarm_cfg_counter.user_data);
+		data->alarm_irq_flag = false;
+	}
+
+	/* Top_value terminal count event */
+	if ((data->top_value_cfg_counter.callback != NULL) &&
+	    ((CYHAL_TIMER_IRQ_TERMINAL_COUNT & event) == CYHAL_TIMER_IRQ_TERMINAL_COUNT)) {
+
+		/* Call User callback for top value */
+		data->top_value_cfg_counter.callback(dev, data->top_value_cfg_counter.user_data);
+	}
+
+	/* NOTE: cyhal handles cleaning of interrupts */
+}
+
+static void ifx_cat1_counter_set_int_pending(const struct device *dev)
+{
+	__ASSERT_NO_MSG(dev != NULL);
+
+	struct ifx_cat1_counter_data *const data = dev->data;
+	const struct ifx_cat1_counter_config *const config = dev->config;
+
+	cyhal_timer_enable_event(&data->counter_obj, CYHAL_TIMER_IRQ_CAPTURE_COMPARE,
+				 config->irq_priority, true);
+	NVIC_SetPendingIRQ((IRQn_Type)config->irqn);
```

Comment:
Would it make more sense to have `irqn` of the right type in the struct and only cast it once on the initial assignment?

Reply:
Changed `config->irqn` to IRQn_TYPE